clang/test/3C/canwrite_constraints.h

```diff
@@ -50,20 +58,23 @@ void unwritable_cast(void((*g)(int *q)) : itype(_Ptr<void(_Ptr<int>)>)) {
 
 // Make sure that FVComponentVariable::equateWithItype prevents both of these
 // from being changed.
-//
-// TODO: Add the correct expected root cause warnings once diagnostic
-// verification is re-enabled
-// (https://github.com/correctcomputation/checkedc-clang/issues/609).
+
+// expected-warning@+1 {{Declaration in non-writable file}}
 void unwritable_func_with_itype(int *p : itype(_Array_ptr<int>)) {}
 
+// expected-warning@+1 {{Declaration in non-writable file}}
 void unwritable_func_with_itype_and_bounds(int *p
                                            : itype(_Array_ptr<int>) count(12)) {
 }
 
 // Test for https://github.com/correctcomputation/checkedc-clang/issues/580
+// expected-warning@+1 {{Declaration in non-writable file}}
 _Itype_for_any(T) void my_generic_function(void *p : itype(_Ptr<T>));
 
 void unwritable_type_argument() {
   int i;
+  // TODO: Why is this warning showing up? Does it relate to an atom
+  // representing the temporary pointer of `&i`?
```

Comment:
Ideally we could avoid adding `Expression in non-writable file` constraints on atoms like this that won't ever be rewritten.  This can cause 3c to over constrain in some edge cases. For example, add a struct declaration ` struct canwrite_struct { int *ptr; };` before the `#include` in `canwrite_constraints.c` and then  add `  struct canwrite_struct foo = {&i};` here. The pointer field is not converted even though it could be.

Reply:
Yes, I was wondering about that kind of problem myself.  Thanks for working out the example; I confirmed it, filed #618, and updated the comment in this PR.  Does the comment look good now?

Reply:
Yes that sounds fine.